### ISSUE TITLE
Better search interface

### ISSUE THREAD

**Issue by vsoch** [NeuroVault/NeuroVault#430]
We need a cohesive, single place that serves the following purposes:

- general search across fields for all statistical maps
- search based on cognitive atlas task (it would be ideal to have concept, but we don't have that level of detail represented in the database
- along with this, a view based on cognitive atlas task, and it would be nice to have a button to easily download all images (for meta analysis, etc.)

This combines the following (old) issues:

#247 
#228 
#217 

It makes sense to first work on the task specific views, and then integrate into a search framework to improve upon what we currently have.

**Comment by chrisfilo** [comment]
+1

**Comment by chrisfilo** [comment]
On the topic of concepts - can we get this via Cognitive Atlas API? A quick call to the API asking for all tasks/contrasts for a given concept? Then query NeuroVault for those tasks/contrasts?

**Comment by vsoch** [comment]
Yes we definitely can. I think it would be more useful to have a concepts view, because people tend to want to understand images in that context. Is it worth making a task / contrast view, or should we just do concept?

**Comment by chrisfilo** [comment]
I would start with Task since it does not require changes to cogatlas API
and provides some useful information. It could be easily expanded to
concepts later (the queryset would just change from asking for one task to
asking for a set of task/contrasts).

**Comment by vsoch** [comment]
ok

**Comment by vsoch** [comment]
Quick update: task view is traditional viewer:

![image](https://cloud.githubusercontent.com/assets/814322/13030642/92359fe6-d266-11e5-86b4-402040f923ac.png)

images are retrieved with json call as with collections, url is like:

      /images/trm_523c7a0a73cf5/task

The second view is to view the graph of the task:

![image](https://cloud.githubusercontent.com/assets/814322/13030664/0900696c-d267-11e5-8396-03a5faebd3c0.png)

This shows the task --> contrasts --> concepts, and I'm going to try and do the following:

- give meaningful colors (and legend)
- provide images as nodes at end, so user can click and go to image page

If there are no images matching the task:

![image](https://cloud.githubusercontent.com/assets/814322/13030653/cc6d116c-d266-11e5-939c-7a8660470045.png)

It might also be nice to have a button to download the image set, and some other random stuffs. Still working!

**Comment by vsoch** [comment]
Colors and legend

![image](https://cloud.githubusercontent.com/assets/814322/13030777/3cf9ebf4-d26b-11e5-8782-70de3aa44f1a.png)

**Comment by vsoch** [comment]
ok, image nodes are added, and clicking an image node goes to the image page. I also added a "guilt trip" note at the bottom listing the images that are not tagged with contrasts (and not included in the tree)

![image](https://cloud.githubusercontent.com/assets/814322/13031008/ac601d96-d271-11e5-8036-0f01b259ab1d.png)

I'm thinking that having to click nodes to expand the tree is not useful - maybe we might have the tree already expanded (the default is to show the task node and contrasts, and user must click to expose concepts and images) and the user is allowed to click a contrast or concept node to download the image set, but images that are not tagged with contrast are of course not included :)

**Comment by vsoch** [comment]
ok I think this looks good for now - we next need to figure out the right way to implement search. @burnash @chrisfilo @rwblair any ideas? It should be efficient, and be able to search across image meta data, along with cognitive atlas tasks (for now) to get to the above view. I'll think about this too.

**Comment by chrisfilo** [comment]
Colors really add a lot to the plot! A few comments:

1) It's not clear where on the hierarchy of the graph "concepts" should live. For example, how would the plot look like if there were multiple concepts associated with one contrast?

2) The bottom line encouraging people to tag contrasts is a great move. It would rephrase it a little bit though, because now is suggests anyone can tag any image. Also I expect that most tasks will not have contrasts tagged so this list can be very long.

3) Expanded tree might be a better default.

4) In terms of search there are multiple ways we can go. Let's first think what would be the usual use cases. I can think of some typical searches:
"Gordon et al. 2015"
"Harvard-Oxford atlas"
"response inhibition"
"motor learning"
"single subject emotion"
"IAPS"
"resting state schizophrenia"

I tested some of those with google (using site:neurovault.org prefix) and results are surprisingly accurate. Maybe we could just embed google search?

Roping in @tyarkoni.

**Comment by vsoch** [comment]
1) Multiple concepts associated with a contrast would produce more concept nodes coming out of the concept (right now there is just a single one)

2) We might have a link to the page, and some way to (through the interface) send a message to the account of the user to ask him/her to tag?

3) Agree

4) Will think more about this.

**Comment by chrisfilo** [comment]
On Sat, Feb 13, 2016 at 7:35 PM, Vanessa Sochat <notifications@github.com>
wrote:

> 1) Multiple concepts associated with a contrast would produce more concept
> nodes coming out of the concept (right now there is just a single one)
>
So this will lead to multiple links to the same NV image on the graph?

> 2) We might have a link to the page, and some way to (through the
> interface) send a message to the account of the user to ask him/her to tag?
>
Seems a bit involved, but we can put it on the TODO list.

**Comment by vsoch** [comment]
Correct, multiple links to the image.

**Comment by vsoch** [comment]
If this becomes an issue with too many images, we can eventually change to just show a node with a count for the number of images, and if this is the case we will likely develop a "contrast specific view." This just isn't an issue now with the limited tagging. It might also be interesting to have a graph of the entire Cognitive Atlas ontology, with nodes labeled based on images tagged to them. Given that NeuroVault holds a somewhat static snapshot of the Cognitive Atlas, we could generate the tree whenever there are changes to the tasks/contrasts and dynamically add old/new images to it (and I say this because querying CogAt to generate it on demand is too time intensive for a dynamic, single view).

**Comment by vsoch** [comment]
Also want to comment that we need more than "just search" - but filtering, and not just of collections, but images and other objects as well. For example, I should be able to filter down collections and/or images based on fields like group, if it's in MNI, thresholded, in addition to the examples @chrisfilo provided. So - perhaps we want an overall search functionality that returns collections and images, and then a dynamic filter to show subsets and sorting of that result.

**Comment by burnash** [comment]
I think Google Custom Search is the easiest way to implement full text search search.

For filtering (or faceted search) different options are available. In the SaaS field, Algolia has a free plan (with limit of 10000 records). Self-hosted options [are plenty](https://www.djangopackages.com/grids/g/search/). [Haystack](http://haystacksearch.org/) seems to be the most popular choice. Among multiple backends Haystack supports Elasticsearch, which I find very flexible and easy to work with.

**Comment by vsoch** [comment]
ok, time to try playing around with haystacks then! Perhaps we will find a needle :)

**Comment by vsoch** [comment]
Is there a reason we can't just search by making API calls, akin to the way to load data into the jquery.datatables? Is Django's native query slower than a third party indexed solution?

**Comment by chrisfilo** [comment]
I'm not sure about the performance of the two solutions.

BTW I found that you can define filtering operation with custom meta tags
using Google Custom Search. Their API is pretty powerfull:
https://developers.google.com/custom-search/docs/overview We could just
expose modality, analysis level, cognitive paragigm through <meta> tags and
let google do the rest. Definitely worth looking into. Advantage of this
solution is that indexing and search would be done by Google, we would have
less code to maintain in the future and we can improve general google
search results pointing to NeuroVault.

**Comment by vsoch** [comment]
That's easy and works fast, but the issue I see is not being able to fine tune a search, and only returning a list of links (not so useful). For example, I should be able to search for not_mni = False, cognitive_paradigm_cogatlas = something, and then return a list of  the images and a button to download (and not just random links to things that I have to click).

I think we have to better describe what we want. If we want a way to intelligently filter collections and images to produce an image set that can be downloaded, then we don't want google search. If we just want to search over words on the site and return a list of pages, then we do.

**Comment by chrisfilo** [comment]
All I'm saying is that we can I think we can get the fine-tuned search you
are talking about using Google. Have a look at:

* https://developers.google.com/custom-search/docs/structured_search
* https://developers.google.com/custom-search/docs/refinements

Some control over the look and feel is also possible:
https://developers.google.com/custom-search/docs/ui

I'm not sure if getting the "download all images returned by this search"
functionality would be possible within this framework. I could not find a
way to get this with their API. It could be possible with some custom
javascript.

**Comment by tyarkoni** [comment]
It seems to me that there are two main use cases here:

a) Someone doing an informal search via the website, who just wants to type queries like the ones @chrisfilo gave examples of. This person doesn't need a structured, complex search form, and if anything, it might put them off. Google should work great for this--almost certainly better than anything else we could come up with.

b) Someone who needs to identify a list of images or collections that meet specific criteria so that they can grab the images or metadata. This person doesn't need a nice web interface, they just need an API and/or nice wrapper around that API (e.g., the default Swagger interface that shows all the parameters and their docs works great). In theory all that's required to make this work is to make sure the API searches over whatever fields you want. I don't know how that would work here, but I imagine it's just a matter of mapping url parameters to accessible fields, and shouldn't be too hard.

I'm not sure I see any need for a unifying search API. It seems like overengineering for now when the alternative is to just tweak the API and write some docs for it.

**Comment by chrisfilo** [comment]
Sounds reasonable. We should probably link to the API docs in the general
search.

**Comment by tyarkoni** [comment]
Oh, to add to that: I think having some compact format for listing results sets of images and collections would be pretty useful. But that could be done on the API side by creating some custom output format that just wraps the returned JSON with an HTML snippet. Then if you query the API from a page on the website, you can just tack on something like ?format=list, which would return HTML instead of JSON.

**Comment by chrisfilo** [comment]
Django Rest Framework provides HTML rendering of API results out of the
box. No extra code is necessary.

**Comment by tyarkoni** [comment]
Yeah but I think what Vanessa meant (or if not, then definitely what I meant) was that we probably want something nicer than just a JSON dump in an HTML page. E.g., a compact representation of each image record where there's a thumbnail of the image on the left, and only the key metadata on the right, with clickable links (e.g., to the collection, tags, user, etc.). Basically something table-like and easily navigable that's intended for humans and not machines.

**Comment by tyarkoni** [comment]
Are there existing API docs, or would new ones need to be written? If the latter, maybe https://github.com/marcgibbons/django-rest-swagger would help? Don't know how good the DRF integration is, but I can vouch for Swagger being easy to add and document. I see #257 is open; is that still on the table? Can't promise I'll have any time for this in the next month or so, but maybe later on...

**Comment by vsoch** [comment]
+1. 

I had also wanted that to improve the current image comparison search, and to integrate the new cognitive atlas view, but this seems like it should be two separate PRs. For this PR, it sounds like what is desired is a wrapper around google search, and to make the Cognitive Atlas views accessible, a simple drop down form for the user to navigate to that. This isn't what I had in mind, but it would be a small step in the right direction to add some kind of search.

**Comment by vsoch** [comment]
Google is so amazing... I feel like a toddler walking through their API docs.

**Comment by vsoch** [comment]
We can add synonyms and Google Analytics! @chrisfilo can you add me to NeuroVault for Analytics?
